### ISSUE TITLE
Setting catalog/seo/category_url_suffix through Admin to an empty string sets value to NULL, which fails return type declaration in code that uses the value

### ISSUE THREAD

**Issue by danemacmillan** [magento/magento2#29442]
### Preconditions (*)

1. `2.3.3-p1`
2. Have products created with the `Generate "category/product" URL Rewrites`config set to `Yes` in Stores->Config->Catalog->SEO.
3. Set `Generate "category/product" URL Rewrites`  to `No` in Stores->Config->Catalog->SEO, which will clear a lot of rows in the `url_rewrite` table, but not all.
4. Ensure there is at least one `url_rewrite.request_path` with a `category/product` URL. That will need to be visited after performing the below.

### Steps to reproduce (*)

1. Use category suffixes, like `.html` in the admin for `catalog/seo/category_url_suffix`.
2. Save.
3. Note that the value in `core_config_data` at path `catalog/seo/category_url_suffix` is a string: `.html`.
4. Empty the value and save, effectively meaning there should be no suffix. 
5. Save.
6. Note that the value in `core_config_data` at path `catalog/seo/category_url_suffix` is of type `NULL`, instead of string.
7. Clear cache. Visit a product page that has or used to have a category path in the URL, using the long URL.

### Expected result (*)

1. Product page should load with a 404.

### Actual result (*)

1. A blank page and 500 HTTP error code loads instead. 
2. See this error instead when displaying errors:

```
 Fatal error: Uncaught TypeError: Return value of Magento\CatalogUrlRewrite\Model\Storage\DynamicStorage::getCategoryUrlSuffix() must be of the type string, null returned in /Users/danemacmillan/projects/linusshops/storefront233/releases/20200729133910803462/app/code/Magento/CatalogUrlRewrite/Model/Storage/DynamicStorage.php:150 
 
 Stack trace: 
 
 #0 /Users/danemacmillan/projects/linusshops/storefront233/releases/20200729133910803462/app/code/Magento/CatalogUrlRewrite/Model/Storage/DynamicStorage.php(171): Magento\CatalogUrlRewrite\Model\Storage\DynamicStorage->getCategoryUrlSuffix('5') 
 #1 /Users/danemacmillan/projects/linusshops/storefront233/releases/20200729133910803462/app/code/Magento/CatalogUrlRewrite/Model/Storage/DynamicStorage.php(102): Magento\CatalogUrlRewrite\Model\Storage\DynamicStorage->findProductRewriteByRequestPath(Array) 
 #2 /Users/danemacmillan/projects/linusshops/storefront233/releases/20200729133910803462/app/code/Magento/UrlRewrite/Model/Storage/AbstractStorage.php(66): Magento\CatalogUrlRewrite\Model\Storag in /Users/danemacmillan/projects/linusshops/storefront233/releases/20200729133910803462/app/code/Magento/CatalogUrlRewrite/Model/Storage/DynamicStorage.php on line 150
```

The reason this happens is due to the declared return type, which only expects and allows `string`, while saving that configuration through the admin to an empty value will actually set `NULL` instead of just an empty string. Regardless of one's ability to reproduce this issue, it is a clear, logical problem.

The offending code and commit, which just randomly declared return types by @slopukhov:

https://github.com/magento/magento2/commit/f70f1b2826280bbe18a73a66dc2ff5f70c8d2d1b#diff-eb1d8c6c0059c0e71a5cbdc544a4558fR144

Insight: simply because return types can be declared does not mean that they *should* be declared. It is not as benign a change as perceived.

### Workaround

Manually go to the `catalog/seo/category_url_suffix` config in the DB and ensure that it contains an empty string instead of `NULL`. Clear the cache, and suddenly those pages will load.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by shikhamis11** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11, here is your [Magento instance](https://i-29442-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-29442-2-4-develop.instances.magento-community.engineering/admin_a1e4](https://i-29442-2-4-develop.instances.magento-community.engineering/admin_a1e4)
Login: `f0ece059` Password: `c5155fd4b63d`
Instance will be terminated in up to 3 hours.

**Comment by shikhamis11** [comment]
hi @danemacmillan 
unfortunately this issue is not reproducible at magento 2.4 instance as all data with category url is cleared after setting 
"category/product" URL Rewrites to `No` in Stores->Config->Catalog->SEO 

Please share additional information if possible to reproduce this issue 

I am keeping this issue open so that if anyone else can reproduce this issue

**Comment by engcom-Bravo** [comment]
Hello @danemacmillan 

Thank you for your report.

We have tried to reproduce the issue on 2.4-develop via the steps provided in the description. The product page is opened without any errors. We are using the URL from `url_rewrite` table, `request_path` column.

Can you please confirm that you are facing the same issue on the latest 2.4-develop? It would help us to verify the issue faster

Thank you in Advance

**Comment by hostep** [comment]
@engcom-Bravo: it looks like this is a duplicate of https://github.com/magento/magento2/issues/24903 which got fixed by https://github.com/magento/magento2/pull/24907 and the fix was released in Magento 2.3.4

So you can probably close this issue 🙂

**Comment by engcom-Bravo** [comment]
Thank you @hostep for the provided information

@danemacmillan We are closing this issue as Fixed on 2.4-develop

Please feel free to comment, create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines)

Thank you
